lib/machine_gun/worker.ex

```diff
@@ -304,4 +309,21 @@ defmodule MachineGun.Worker do
 
   def parse_reason({:shutdown, reason}), do: %Error{reason: reason}
   def parse_reason(reason), do: %Error{reason: reason}
+
+  defp last_request?(headers) do
+    Enum.any?(headers, fn
+      {"connection", v} -> v == "close"
+      _ -> false
+    end)
+  end
+
+  defp close_connection(%Worker{gun_pid: gun_pid, gun_ref: gun_ref} = worker) do
+    :ok = :gun.close(gun_pid)
+    true = :erlang.demonitor(gun_ref, [:flush])
+
+    worker
+    |> Map.put(:gun_pid, nil)
+    |> Map.put(:gun_ref, nil)
+    |> reply_error(:normal)
```

Comment:
Looks good! Let's use :closed as reason, also please update the structure with | to let the compile check its fields:

```
%{worker | gun_pid: nil, gun_ref: nil} |> reply_error(:closed)
```